index.rst

```diff
@@ -58,24 +58,316 @@ We describe the design and implementation of the LSST Alert Distribution System,
+Having a group of filters share a copy of the full alert stream
+minimizes the number of full streams and the amount of data
+that needs to be pulled from Kafka, which is necessary given
+a cap on the network bandwidth available for the system.
+Additionally, the load of a large number (~10) of consumers,
+i.e. a large number of full streams, may cause filters to lag
+and not be able to keep up with the volume and real-time velocity of alerts.
+(See DMTN-028.)
+
+The previous design deploys filters in groups in order to avoid the
+bandwidth needed to otherwise support the design of a one-stream-per-filter
+(and therefore 100+ streams) on a single Kafka instance.
+The major drawback in this deployment is that groups of filters are running
+within a single container and within a single executed Python script.
+Filters cannot be isolated to their own environments, and one filter
+may slow or otherwise affect another.
+
+The design described here attempts to avoid both the bandwidth issue
+and the issue of shared filter environments by placing
+components methodically on the hardware available to the system.
+This design separates each filter into its own Docker container so that
+filters should have no effect on each other.
+However, instead of having filter containers deployed on potentially
+many separate nodes all consuming streams from one Kafka instance running
+on its own dedicated node, a system of downstream Kafka mirrors are deployed
+on several nodes from which co-located filters can read streams locally,
+cutting down on the data throughput between nodes.
+Filters run in separate Docker containers which are placed on the
+same node as a Kafka instance mirroring the full stream pulled from the
+upstream central Kafka hub where alerts are sent.
+The number of mirrored Kafka instances / groups of filters able to
+be supported is then determined by the node-to-node available bandwidth,
+but the filters reading from the downstream local Kafka instances
+can take advantage of the internal bandwidth of the node on which they run.
+
+The updated deployment plan of the alert distribution system
+including filtering with the mini-broker is shown in the figure below.
+Content of the alerts is created in science pipelines, potentially
+parallelized by CCD.
+In the alert_stream repo, AlertProducers serialize Python dicts
+into Avro format when alerts are sent to a central Kafka broker,
+or cluster of Kafka brokers acting as one unit.
+For testing, a single AlertProducer can be deployed from the
+alert_stream repo that sends local Avro alert visit files to Kafka.
+Each Kafka broker in this central hub is deployed on its own
+node either with Zookeeper on the same node or Zookeeper on a separate
+node with no other services running.
+The main Kafka system streams to and feeds downstream community broker
+consumers and sinks to the alert database.
+
+
+.. figure:: deployment-diagram.png
```

Comment:
From the figure, it looks as though the outputs of the filters on node `N` are only available to other filters running on the same node, rather than going back to the central node. Is that correct? I don't think it's mentioned in the text (or did I miss it?).

Reply:
Yes, I should have put that in.  So consumers that pull directly from Kafka need to use the instance where that filter is running.  That's how I wrote all the scripts.  Or all filtered streams need to be directed/sinked to wherever end users look to grab their data.